### ISSUE TITLE
Deteriorate in performance of latest libtorrent

### ISSUE THREAD

**Issue by lethanhviet** [arvidn/libtorrent#509]
Hi,

After able to build the new libtorrent from https://github.com/arvidn/libtorrent/issues/500, I noticed deluge/libtorrent has performed worse than before. Always high CPU usage. Take very long time to connect to swarm. Connection with peers just drop randomly. I tried to drop back to the libtorrent version that comes with debian Jessie (0.16.8.0) and notice a much better overall performance.

Sorry just cant find the log even though I built libtorrent with ./configure --enable-python-binding --with-libiconv --enable-logging=default --enable-debug=yes

**Comment by arvidn** [comment]
> Always high CPU usage.

Do you know where it spends the CPU? (say, with a CPU profile)

**Comment by arvidn** [comment]
the debug log is reported as alerts and has to be enabled by the alert mask

**Comment by arvidn** [comment]
as for the time it takes to join a swarm, what's your connect_speed set to? it defaults to 6 right now (that's 6 connection attempts per second) which may be a bit low. Previously this limit was specified as half-open limit (based on a windows XP era restriction). Now it's specified in connection attempts per second. In practice 6 per second could be lower than you saw with the previous limit (but it's also possible for it to effectively be a higher limit, depending on your connection and connect success rate and latency)

**Comment by lethanhviet** [comment]
my current connection_speed is set at 500. I have rebuild libtorrent with --enable-logging=yes. Please see screenshot of atop. Deluge has 56 torrents seeding with only 2 3 are on active. If with libtorrent 0.16.8.0 it's only around 4-5%.
![screenshot](https://cloud.githubusercontent.com/assets/12855116/13402213/e5e59802-df41-11e5-93d4-ad942342cde8.png)

**Comment by arvidn** [comment]
could you take a profile of deluged (with a release build with debug symbols of libtorrent)? I've had good luck with "prof" on linux for this.

Are you confident the logging itself isn't what's causing the CPU usage? The amount of logging (if enabled) can be quite substantial. Perhaps deluge enables all alerts by default for instance (which would make it get all peer logs)

**Comment by arvidn** [comment]
@cas-- does deluge enable all alerts by any chance? libtorrent 1.1 has transitioned all debug logging over to alerts (that are turned off by default in the alert mask). If deluge would enable all bits there would likely be a lot of string copying and logging going on, probably pegging a core

**Comment by arvidn** [comment]
as for peer connections dropping. how do you measure that, and compare against a previous version of libtorrent? would it be possible to quantify it?

I've spent some time reviewing logs of short lived connections closed by the other end. I haven't been able to establish any pattern to determine what might cause them.

**Comment by cas--** [comment]
@arvidn We do use a mask: https://github.com/deluge-torrent/deluge/blob/develop/deluge/core/alertmanager.py#L41

**Comment by lethanhviet** [comment]
after rebuilding libtorrent with --disable-debug --enable-logging=no deluge/libtorrent's cpu usage has been much better. Before, with libtorrent built with --enable-loging=yes, deluge is always started with -l warning but as I check deluge's log, I see deluge reported no issue.

But with the new config, I see deluge/libtorrent has crashed a few time just the last 2 days.

As for the peer connections, it's just me watching the performance of deluge before and now.

**Comment by arvidn** [comment]
I don't know what "-l warning" does in deluge. You would expect it to perhaps log some hints of why peers disconnect? or why it's using a lot of CPU?

Is it possible that it's the debug-build that's causing high CPU usage? rather than logging. There are definitely a lot of somewhat expensive invariant checks that are enabled by default in debug builds, to verify internal consistency. If the CPU issue goes away in release builds, I think it's probably a good idea to consider that issue solved.

The other issue you mention is peers dropping, it taking longer to join a swarm and worse overall performance. Does "performance" refer to the CPU issue or download rate?
Is it likely the dropping-peers and slow-swarm-join are in fact caused by the same underlying problem? (I would imagine peers dropping randomly to make it take longer to fill ones connections slots).

joining a swarm btw, just to be precise, you mean to reach 50 peer connections? (or whatever limit you may have)

**Comment by arvidn** [comment]
@lethanhviet my testing seems to suggest that uTP connections are especially prone to be disconnected. Is this consistent with what you experience? If you disable uTP (and just use TCP) does it improve things?

**Comment by arvidn** [comment]
@lethanhviet I see the issue you describe with dropping connections when running over a VPN. I've tracked it down so far to the fact that this network does not support IP fragmentation. uTP relies on IP fragmentation to work for its Path MTU discovery probe packets.

In this case, whenever a packet larger than the MTU is sent (regardless of whether it's a probe or not) will inevitably break the connection. It will never be delivered, regardless of resends. Under normal circumstances this would result in lowering our MTU limits (and continue the binary search of the PMTU).

As far as I can tell, the reason it's much more likely to break in 1.1 than 1.0 is that I optimized the network protocol to requirer fewer round-trips before transfers can start. This means bittorrent messages are coalesced into larger packets and more quickly sends a large enough packet to exceed the limit.

I imagine 1.0 will simply have a tendency to not upload under these conditions, since any peer it starts uploading to (at any significant rate) would be disconnected as the packets become too large.

I will try to mitigate this by keeping state of packet sizes and whether they were dropped or ACKed. That way I can have a heuristic where new sockets will use increasingly conservative PMTU initial values/limits.

If you would like to confirm or rule out that this is what you're experiencing, try this patch:

```
diff --git a/src/utp_stream.cpp b/src/utp_stream.cpp
index 318bd89..2eb1a52 100644
--- a/src/utp_stream.cpp
+++ b/src/utp_stream.cpp
@@ -2665,5 +2675,5 @@ void utp_socket_impl::init_mtu(int link_mtu, int utp_mtu)
        // set the ceiling to what we found out from the interface
-       m_mtu_ceiling = utp_mtu;
+       m_mtu_ceiling = 1000;
  
        // however, start the search from a more conservative MTU
        int overhead = link_mtu - utp_mtu;
```
